src/clojure/zero_one/geni/arrow.clj

```diff
@@ -0,0 +1,222 @@
+               :get (if (.isNullAt (row row-info) (idx2 row-info)) nil (.getBoolean (row row-info) (idx2 row-info))))
+
+    :date
+    (case action
+      :set (.set ^TimeStampMilliVector (vector value-info) (idx1 value-info) (.getTime ^java.sql.Date (value value-info)))
+      :make-vector (TimeStampMilliVector. name allocator)
+      :get (if (.isNullAt (row row-info) (idx2 row-info)) nil (.getDate (row row-info) (idx2 row-info)))))
+  )
+
+
+(defn typed-set [v  idx value type]
+  (typed-action :set type (->ValueInfo v idx value) nil nil nil))
+
+(defn typed-make-vector [name allocator type]
+  (typed-action :make-vector type nil nil name allocator))
+
+(defn typed-get [row idx type]
+  (typed-action :get type nil (->RowInfo row idx) nil nil))
+
+
+(defn schema->clojure [^Schema schema]
+  (map
+   #(hash-map
+     :type %1
+     :name %2)
+   (->> schema .fields (map #(keyword (.. % dataType typeName))))
+   (->> schema .fields (map #(.. % name)))))
+
+(defn set-null-or-value [v ^long idx value type]
+  (if (nil? value)
+    (.setNull v idx)
+    (typed-set v idx value type)))
+
+(defn fill-vector [vector values type]
+  (let [_ (.allocateNew vector (count values))
+        idx-vals (map #(hash-map :idx %1 :value %2)
+                      (range)
+                      values)
+        _ (run! #(set-null-or-value vector (:idx  %) (:value %) type)
+                idx-vals )
+        _ (.setValueCount ^ValueVector vector (count values))]
+
+
+    vector))
+
+
+(defn rows->data [rows schema-maps]
+  (partition
+   (count schema-maps)
+   (for [row rows field-idx (range (count schema-maps))]
+     (let [type (:type (nth schema-maps field-idx))]
+       (typed-get row field-idx type)))))
+
+
+(defn rows->vectors [rows schema-maps]
+(let [allocator (RootAllocator. Long/MAX_VALUE)
+          schema-names (map :name schema-maps)
+          data (rows->data rows schema-maps)
+          transposed-data (apply pmap list data)
+          vectors (pmap
+                   #(fill-vector
+                     (typed-make-vector
+                      (:name %1)
+                      allocator
+                      (:type % 1))
+                     %2
+                     (:type %1))
+                   schema-maps
+                   transposed-data)
+          ]
+      vectors
+      )
+
+  )
+
+(defn export-rows! [rows schema-maps out-dir]
+  (when (pos? (count rows))
+    (let [vectors (rows->vectors rows schema-maps)
+          vector-fields (map #(.getField ^ValueVector %) vectors)
+          root (VectorSchemaRoot. vector-fields vectors)
+          out-file (java.io.File/createTempFile "geni" ".ipc" (io/file out-dir))]
+
+      (with-open [out (Channels/newChannel  (clojure.java.io/output-stream out-file))
+                  writer  (ArrowStreamWriter. root nil out)
+                  ]
+        (doto writer
+          (.start)
+          (.writeBatch)
+          (.end)))
+
+      (.getPath out-file))
+
+    ))
+
+(defn collect-to-arrow
```

Comment:
I believe this is the only function we want in `core`. Should we make all the other functions private?

Reply:
done